runtime/near-vm-logic/src/gas_counter.rs

```diff
@@ -65,32 +96,65 @@ impl GasCounter {
+        let new_used_gas =
+            new_burnt_gas.checked_add(new_promises_gas).ok_or(HostError::IntegerOverflow)?;
         if new_burnt_gas <= self.max_gas_burnt && (self.is_view || new_used_gas <= self.prepaid_gas)
         {
-            self.burnt_gas = new_burnt_gas;
-            self.used_gas = new_used_gas;
+            use std::cmp::min;
+            if promise_gas != 0 && !self.is_view {
+                self.fast_counter.gas_limit =
+                    min(self.max_gas_burnt, self.prepaid_gas - new_promises_gas);
+            }
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            self.promises_gas = new_promises_gas;
             Ok(())
         } else {
-            use std::cmp::min;
-            let res = if new_burnt_gas > self.max_gas_burnt {
-                Err(HostError::GasLimitExceeded.into())
-            } else if new_used_gas > self.prepaid_gas {
-                Err(HostError::GasExceeded.into())
-            } else {
-                unreachable!()
-            };
-
-            let max_burnt_gas = min(self.max_gas_burnt, self.prepaid_gas);
-            self.burnt_gas = min(new_burnt_gas, max_burnt_gas);
-            self.used_gas = min(new_used_gas, self.prepaid_gas);
-
-            res
+            Err(self.process_gas_limit(new_burnt_gas, new_used_gas))
         }
     }
 
+    // Optimized version of above function for cases where no promises involved.
+    pub fn burn_gas(&mut self, value: Gas) -> Result<()> {
+        let new_burnt_gas =
+            self.fast_counter.burnt_gas.checked_add(value).ok_or(HostError::IntegerOverflow)?;
+        if new_burnt_gas <= self.fast_counter.gas_limit {
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            Ok(())
+        } else {
+            Err(self.process_gas_limit(new_burnt_gas, new_burnt_gas + self.promises_gas))
+        }
+    }
+
+    fn process_gas_limit(&mut self, new_burnt_gas: Gas, new_used_gas: Gas) -> VMLogicError {
+        use std::cmp::min;
+        if new_used_gas > self.prepaid_gas {
+            // Ensure that contract never burn more than `max_gas_burnt`, even if paid for more.
+            let hard_burnt_limit = min(self.prepaid_gas, self.max_gas_burnt);
+            self.fast_counter.burnt_gas = min(new_burnt_gas, hard_burnt_limit);
+            // Technically we shall do `self.promises_gas = 0;` or error paths, as in this case
+            // no promises will be kept, but that would mean protocol change.
+            // See https://github.com/near/nearcore/issues/5148.
+            // TODO: consider making this change!
+            assert!(self.prepaid_gas >= self.fast_counter.burnt_gas);
+            self.promises_gas = self.prepaid_gas - self.fast_counter.burnt_gas;
+            HostError::GasExceeded.into()
+        } else {
+            self.fast_counter.burnt_gas = min(new_burnt_gas, self.max_gas_burnt);
+            self.promises_gas = 0;
```

Comment:
Hm, if I replace `self.promises_gas = 0;` with `panic!()`, `cargo t -p near-vm-runner -p near-vm-logic` passes successfully for me. Ie, we don't hit this code path in test *at all*. This doesn't inspire confidence, let me see if I can add one more test here.

Reply:
Fixed, although TBH do not like the code I have to write there. IMO, promises gas shall be zero'ed out instead.